src/jit/codegenxarch.cpp

```diff
@@ -7366,6 +7366,66 @@ void CodeGen::genSSE2BitwiseOp(GenTreePtr treeNode)
     inst_RV_RV(ins, targetReg, operandReg, targetType);
 }
 
+//-----------------------------------------------------------------------------------------
+// genSSE41RoundOp - generate SSE41 code for the given tree as a round operation
+//
+// Arguments:
+//    treeNode  - tree node
+//
+// Return value:
+//    None
+//
+// Assumptions:
+//     i) SSE4.1 is supported by the underlying hardware
+//    ii) tree oper is a GT_INTRINSIC
+//   iii) tree type is floating point type
+//    iv) tree oper is CORINFO_INTRINSIC_Round, _Ceiling, or _Floor
+//     v) caller of this routine needs to call genProduceReg()
+void CodeGen::genSSE41RoundOp(GenTree* treeNode)
+{
+    // i) SSE4.1 is supported by the underlying hardware
+    assert(compiler->compSupports(InstructionSet_SSE41));
+
+    // ii) tree oper is a GT_INTRINSIC
+    assert(treeNode->OperGet() == GT_INTRINSIC);
+
+    GenTree*    srcNode  = treeNode->gtGetOp1();
+
+    instruction ins      = ins_FloatRound(treeNode->TypeGet());
+    emitAttr    size     = emitTypeSize(treeNode);
+
+    regNumber   dstReg   = treeNode->gtRegNum;
+    regNumber   srcReg   = srcNode->gtRegNum;
+
+    // iii) tree type is floating point type
+    assert(varTypeIsFloating(srcNode));
+    assert(srcNode->TypeGet() == treeNode->TypeGet());
+
+    genConsumeOperands(treeNode->AsOp());
```

Comment:
The `treeNode` parameter should be `GenTreeOp*` to begin with.

Reply:
Will fix. The `CORINFO_INTRINSIC_Sqrt` function should probably be updated as well (in a separate PR)

Reply:
Actually, it seems like most functions take a `GenTreePtr` and cast to `GenTreeOp` via `->AsOp()` only when required.

Reply:
(This is looking in the codegen layer, and at similar things like `genSSE2BitwiseOp`)

Reply:
> Actually, it seems like most functions take a GenTreePtr and cast to GenTreeOp via ->AsOp() only when required.

Yes, that's unfortunately the case in existing code. Code that has been modified or added recently tends not to use this "pattern".

Reply:
Fixed.